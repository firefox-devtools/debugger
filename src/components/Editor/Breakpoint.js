// @flow
import React, { Component } from "react";
import { isEnabled } from "devtools-config";
import ReactDOM from "react-dom";

import classnames from "classnames";
import Svg from "../shared/Svg";

import { getDocument, showSourceText, toEditorLine } from "../../utils/editor";

const breakpointSvg = document.createElement("div");
ReactDOM.render(<Svg name="breakpoint" />, breakpointSvg);

function makeMarker(isDisabled: boolean) {
  const bp = breakpointSvg.cloneNode(true);
  bp.className = classnames("editor new-breakpoint", {
    "breakpoint-disabled": isDisabled,
    "folding-enabled": isEnabled("codeFolding")
  });

  return bp;
}

class Breakpoint extends Component {
  props: {
    breakpoint: Object,
    selectedSource: Object,
    editor: Object
  };

  addBreakpoint: Function;

  constructor() {
    super();
    this.addBreakpoint = this.addBreakpoint.bind(this);
  }

  addBreakpoint() {
    const { breakpoint, editor, selectedSource } = this.props;

    // Hidden Breakpoints are never rendered on the client
    if (breakpoint.hidden) {
      return;
    }

    // NOTE: we need to wait for the breakpoint to be loaded
    // to get the generated location
    if (!selectedSource || breakpoint.loading) {
      return;
    }

    const sourceId = selectedSource.get("id");
    const line = toEditorLine(sourceId, breakpoint.location.line);

    showSourceText(editor, selectedSource.toJS());

    editor.codeMirror.setGutterMarker(
      line,
      "breakpoints",
      makeMarker(breakpoint.disabled)
    );

    editor.codeMirror.addLineClass(line, "line", "new-breakpoint");
    if (breakpoint.condition) {
      editor.codeMirror.addLineClass(line, "line", "has-condition");
    } else {
      editor.codeMirror.removeLineClass(line, "line", "has-condition");
    }
  }

  shouldComponentUpdate(nextProps: any) {
    const { editor, breakpoint, selectedSource } = this.props;
    return (
      editor !== nextProps.editor ||
      breakpoint.disabled !== nextProps.breakpoint.disabled ||
      breakpoint.hidden !== nextProps.breakpoint.hidden ||
      breakpoint.condition !== nextProps.breakpoint.condition ||
      breakpoint.loading !== nextProps.breakpoint.loading ||
      selectedSource !== nextProps.selectedSource
    );
  }

  componentDidMount() {
    this.addBreakpoint();
  }

  componentDidUpdate() {
    this.addBreakpoint();
  }

  componentWillUnmount() {
    const { editor, breakpoint, selectedSource } = this.props;

    if (!selectedSource) {
      return;
    }

    if (breakpoint.loading) {
      return;
    }

    const sourceId = selectedSource.get("id");
    const doc = getDocument(sourceId);
    if (!doc) {
      return;
    }

    const line = toEditorLine(sourceId, breakpoint.location.line);

    // NOTE: when we upgrade codemirror we can use `doc.setGutterMarker`
    if (doc.setGutterMarker) {
      doc.setGutterMarker(line, "breakpoints", null);
    } else {
      editor.codeMirror.setGutterMarker(line, "breakpoints", null);
    }

    doc.removeLineClass(line, "line", "new-breakpoint");
    doc.removeLineClass(line, "line", "has-condition");
  }

  render() {
    return null;
  }
}

export default Breakpoint;
