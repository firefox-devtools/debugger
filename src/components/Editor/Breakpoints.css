/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at <http://mozilla.org/MPL/2.0/>. */

.theme-light {
  --gutter-hover-background-color: #dde1e4;

  --breakpoint-fill: var(--blue-50);
  --breakpoint-stroke: var(--blue-60);

  --breakpoint-disabled-fill: var(--blue-55);
  --breakpoint-disabled-stroke: var(--blue-55);
}

.theme-dark {
  --gutter-hover-background-color: #414141;

  --breakpoint-fill: var(--blue-55);
  --breakpoint-stroke: var(--blue-40);

  --breakpoint-disabled-fill: var(--blue-55);
  --breakpoint-disabled-stroke: var(--blue-55);
}

.theme-light,
.theme-dark {
  --logpoint-fill: var(--theme-graphs-purple);
  --logpoint-stroke: var(--purple-60);
  --breakpoint-condition-fill: var(--theme-graphs-yellow);
  --breakpoint-condition-stroke: var(--theme-graphs-orange);
  --breakpoint-inactive-opacity: 0.3;
  --breakpoint-disabled-opacity: 0.6;
}

/* Standard gutter breakpoints */
.editor-wrapper .breakpoints {
  position: absolute;
  top: 0;
  left: 0;
}

.new-breakpoint .CodeMirror-linenumber {
  pointer-events: none;
}

:not(.empty-line):not(.new-breakpoint)
  > .CodeMirror-gutter-wrapper:hover
  > .CodeMirror-linenumber::after {
  content: "";
  position: absolute;
  /* paint below the number */
  z-index: -1;
  top: 0;
  left: 0;
  right: -7px;
  bottom: 0;
  height: 15px;
  background-color: var(--gutter-hover-background-color);
  mask: url(/images/breakpoint.svg) no-repeat;
  mask-size: auto 15px;
  mask-position: right;
}

.editor.new-breakpoint svg {
  fill: var(--breakpoint-fill);
  stroke: var(--breakpoint-stroke);
  width: 60px;
  height: 15px;
  position: absolute;
  top: 0px;
  right: -4px;
}

.editor .breakpoint {
  position: absolute;
  right: -2px;
}

.editor.new-breakpoint.folding-enabled svg {
  right: -16px;
}

.new-breakpoint.has-condition .CodeMirror-gutter-wrapper svg {
  fill: var(--breakpoint-condition-fill);
  stroke: var(--breakpoint-condition-stroke);
}

.new-breakpoint.has-log .CodeMirror-gutter-wrapper svg {
  fill: var(--logpoint-fill);
  stroke: var(--logpoint-stroke);
}

.editor.new-breakpoint.breakpoint-disabled svg {
  fill-opacity: var(--breakpoint-disabled-opacity);
  stroke-opacity: var(--breakpoint-disabled-opacity);
}

/* Columnn breakpoints */
.column-breakpoint {
  display: inline;
  padding-inline-start: 1px;
  padding-inline-end: 1px;
}

.column-breakpoint:hover {
  background-color: transparent;
}

.column-breakpoint svg {
  display: inline-block;
  cursor: pointer;
  height: 13px;
  width: 11px;
  vertical-align: top;
  fill: var(--breakpoint-fill);
  stroke: var(--breakpoint-stroke);
  fill-opacity: var(--breakpoint-inactive-opacity);
  stroke-opacity: var(--breakpoint-inactive-opacity);
}

.column-breakpoint.active svg {
  fill: var(--breakpoint-fill);
  stroke: var(--breakpoint-stroke);
  fill-opacity: 1;
  stroke-opacity: 1;
}

.column-breakpoint.disabled svg {
  fill-opacity: var(--breakpoint-disabled-opacity);
  stroke-opacity: var(--breakpoint-disabled-opacity);
}

.column-breakpoint.has-log.disabled svg {
  fill-opacity: 0.5;
  stroke-opacity: 0.5;
}

.column-breakpoint.disabled:not(.has-condition):not(.has-log) svg {
  fill: var(--breakpoint-disabled-fill);
  stroke: var(--breakpoint-disabled-stroke);
}

.column-breakpoint.has-condition svg {
  fill: var(--breakpoint-condition-fill);
  stroke: var(--breakpoint-condition-stroke);
}

.column-breakpoint.has-log svg {
  fill: var(--logpoint-fill);
  stroke: var(--logpoint-stroke);
}

.img.column-marker {
  background-image: url(/images/column-marker.svg);
}
