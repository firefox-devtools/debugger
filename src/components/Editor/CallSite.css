.call-site {
  background: #f0f9ff;
  cursor: pointer;
  position: relative;
}

.call-site::before {
  content: "";
  position: absolute;
  width: 100%;
  height: calc(100% - 2px);
  border-bottom: 2px solid #aed3ef;
}

.call-site-bp {
  cursor: pointer;
  position: relative;
}

.debug-expression.call-site-bp,
.call-site-bp {
  background-color: #fce7e7;
}

.call-site-bp::before {
  content: "";
  position: absolute;
  width: 100%;
  height: calc(100% - 2px);
  border-bottom: 2px solid red;
}

.theme-dark .call-site {
  background-color: #4b5462;
}

.theme-dark .call-site::before {
  border-bottom-color: #5f78a4;
}

.theme-dark .call-site-bp {
  background-color: #4b3f3f;
}

.theme-dark .call-site-bp::before {
  border-bottom-color: #dd4d4d;
}
