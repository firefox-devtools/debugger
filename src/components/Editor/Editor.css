.editor-wrapper {
  --debug-line-border: rgb(145, 188, 219);
  --debug-expression-background: rgba(202, 227, 255, 0.5);
  --editor-searchbar-height: 27px;
  --editor-second-searchbar-height: 27px;
}

.theme-dark .editor-wrapper {
  --debug-expression-background: #54617e;
  --debug-line-border: #7786a2;
}

.editor-wrapper .CodeMirror-linewidget {
  margin-right: -7px;
}

.theme-dark {
  --theme-conditional-breakpoint-color: #9fa4a9;
}

.theme-light {
  --theme-conditional-breakpoint-color: var(--theme-body-color);
}

.paused .in-scope .CodeMirror-line,
.paused .in-scope .CodeMirror-linenumber {
  opacity: 1;
}

.paused .CodeMirror-line,
.paused .CodeMirror-linenumber {
  opacity: 0.7;
}

/**
 * There's a known codemirror flex issue with chrome that this addresses.
 * BUG https://github.com/devtools-html/debugger.html/issues/63
 */
.editor-wrapper {
  position: absolute;
  height: calc(100% - 31px);
  width: calc(100% - 1.5px);
  top: 30px;
  left: 0px;
  --editor-footer-height: 24px;
}

html[dir="rtl"] .editor-mount {
  direction: ltr;
}

.editor-wrapper .breakpoints {
  position: absolute;
  top: 0;
  left: 0;
}

.function-search {
  max-height: 300px;
  overflow: hidden;
}

.function-search .results {
  height: auto;
}

.editor.hit-marker {
  height: 14px;
}

.editor-wrapper .highlight-lines {
  background: var(--theme-selection-background-hover);
}

.coverage-on .CodeMirror-code :not(.hit-marker) .CodeMirror-line,
.coverage-on .CodeMirror-code :not(.hit-marker) .CodeMirror-gutter-wrapper {
  opacity: 0.5;
}

.editor.new-breakpoint svg {
  fill: var(--theme-selection-background);
  width: 60px;
  height: 14px;
  position: absolute;
  top: 0px;
  right: -4px;
}

.inline-bp {
  background-color: #9ddfff;
  width: 20px;
  padding: 0px 5px;
  margin: 0px 4px;
  border-radius: 5px;
  border-color: blue;
  border: 1px solid #00b6ff;
}

.editor.new-breakpoint.folding-enabled svg {
  right: -16px;
}

.new-breakpoint.has-condition svg {
  fill: var(--theme-graphs-yellow);
}

.editor.new-breakpoint.breakpoint-disabled svg {
  opacity: 0.3;
}

.editor.column-breakpoint svg {
  fill: var(--theme-selection-background);
  vertical-align: middle;
  width: 17px;
  height: 14px;
}

.editor.column-breakpoint.breakpoint-disabled svg {
  opacity: 0.3;
}

.CodeMirror {
  width: 100%;
  height: 100%;
}

.editor-wrapper .editor-mount {
  width: 100%;
  background-color: var(--theme-body-background);
}

.CodeMirror-linenumber {
  font-size: 11px;
  line-height: 14px;
}

.folding-enabled .CodeMirror-linenumber {
  text-align: left;
  padding: 0 0 0 2px;
}

/* set the linenumber white when there is a breakpoint */
.new-breakpoint .CodeMirror-gutter-wrapper .CodeMirror-linenumber {
  color: white;
}

/* move the breakpoint below the other gutter elements */
.new-breakpoint .CodeMirror-gutter-elt:nth-child(2) {
  z-index: 0;
}

.editor-wrapper .CodeMirror-line {
  font-size: 11px;
}

.theme-dark .editor-wrapper .CodeMirror-line .cm-comment {
  color: var(--theme-content-color3);
}

.debug-expression {
  background-color: var(--debug-expression-background);
}

.new-debug-line .CodeMirror-line {
  background-color: transparent !important;
  outline: var(--debug-line-border) solid 1px;
}

/* Don't display the highlight color since the debug line
   is already highlighted */
.new-debug-line .CodeMirror-activeline-background {
  display: none;
}

.highlight-line .CodeMirror-line {
  animation: fade-highlight-out 1.5s normal forwards;
}

@keyframes fade-highlight-out {
  0% {
    background-color: var(--theme-highlight-gray);
  }
  100% {
    background-color: transparent;
  }
}

.theme-dark .highlight-line .CodeMirror-line {
  animation: fade-highlight-out-dark 1.5s normal forwards;
}

@keyframes fade-highlight-out-dark {
  0% {
    background-color: var(--theme-content-color3);
  }
  100% {
    background-color: transparent;
  }
}

.CodeMirror-guttermarker-subtle {
  visibility: hidden;
}

.visible {
  visibility: visible;
}
