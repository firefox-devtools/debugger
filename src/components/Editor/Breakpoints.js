/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at <http://mozilla.org/MPL/2.0/>. */

// @flow
import { connect } from "react-redux";
import React, { Component } from "react";

import Breakpoint from "./Breakpoint";

import { getSelectedSource } from "../../selectors";
import getVisibleBreakpoints from "../../selectors/visibleBreakpoints";
import { makeLocationId } from "../../utils/breakpoint";
import { isLoaded } from "../../utils/source";

import type { SourceRecord, BreakpointsMap } from "../../reducers/types";

type Props = {
  selectedSource: SourceRecord,
  breakpoints: BreakpointsMap,
  editor: Object
};

class Breakpoints extends Component<Props> {
  shouldComponentUpdate(nextProps: Props) {
    if (
      nextProps.selectedSource &&
      !isLoaded(nextProps.selectedSource.toJS())
    ) {
      return false;
    }

    return true;
  }

  render() {
    const { breakpoints, selectedSource, editor } = this.props;

    if (!selectedSource || !breakpoints || selectedSource.get("isBlackBoxed")) {
      return null;
    }

    return (
      <div>
        {breakpoints.valueSeq().map(bp => {
          return (
            <Breakpoint
              key={makeLocationId(bp.location)}
              breakpoint={bp}
              selectedSource={selectedSource}
              editor={editor}
            />
          );
        })}
      </div>
    );
  }
}

export default connect(
  state => ({
    breakpoints: getVisibleBreakpoints(state),
    selectedSource: getSelectedSource(state)
  }),
)(Breakpoints);
