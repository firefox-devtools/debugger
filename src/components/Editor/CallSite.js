// @flow
import { Component } from "react";

import { markText, toEditorRange } from "../../utils/editor";
require("./CallSite.css");

type MarkerType = {
  clear: Function
};

type Props = {
  callSite: Object,
  editor: Object,
  source: Object,
  breakpoint: Object,
  showCallSite: Boolean
};
export default class CallSite extends Component<Props> {
  addCallSite: Function;
  marker: ?MarkerType;

  constructor() {
    super();

    this.marker = undefined;
    const self: any = this;
    self.addCallSite = this.addCallSite.bind(this);
    self.clearCallSite = this.clearCallSite.bind(this);
  }

  addCallSite(nextProps: Props) {
    const { editor, callSite, breakpoint, source } = nextProps || this.props;
    const className = !breakpoint ? "call-site" : "call-site-bp";
    const sourceId = source.get("id");
    const editorRange = toEditorRange(sourceId, callSite.location);
    this.marker = markText(editor, className, editorRange);
  }

  clearCallSite() {
    if (this.marker) {
      this.marker.clear();
      this.marker = null;
    }
  }

  shouldComponentUpdate(nextProps: any) {
    return this.props.editor !== nextProps.editor;
  }

  componentDidMount() {
    const { breakpoint, showCallSite } = this.props;

    if (!breakpoint && !showCallSite) {
      return;
    }

    this.addCallSite();
  }

  componentWillReceiveProps(nextProps: Props) {
    const { breakpoint, showCallSite } = this.props;

    if (nextProps.breakpoint !== breakpoint) {
      if (this.marker) {
        this.marker.clear();
      }
      this.addCallSite(nextProps);
    }

    if (nextProps.showCallSite !== showCallSite) {
      if (nextProps.showCallSite) {
        if (!this.marker) {
          this.addCallSite();
        }
      } else if (!nextProps.breakpoint) {
        this.clearCallSite();
      }
    }
  }

  componentWillUnmount() {
    if (!this.marker) {
      return;
    }
    this.marker.clear();
  }

  render() {
    return null;
  }
}

CallSite.displayName = "CallSite";
