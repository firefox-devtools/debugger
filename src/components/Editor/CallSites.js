/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at <http://mozilla.org/MPL/2.0/>. */

import React, { Component } from "react";
import { connect } from "react-redux";
import { features } from "../../utils/prefs";

import { range, keyBy, isEqualWith } from "lodash";

import CallSite from "./CallSite";

import {
  getSelectedSource,
  getSymbols,
  getSelectedLocation,
  getBreakpointsForSource
} from "../../selectors";

import { getTokenLocation, isWasm } from "../../utils/editor";

import actions from "../../actions";

function getCallSiteAtLocation(callSites, location) {
  return callSites.find(callSite =>
    isEqualWith(callSite.location, location, (cloc, loc) => {
      return (
        loc.line === cloc.start.line &&
        (loc.column >= cloc.start.column && loc.column <= cloc.end.column)
      );
    })
  );
}

class CallSites extends Component {
  props: {
    symbols: Array<Symbol>,
    callSites: Array<Symbol>,
    editor: Object,
    breakpoints: Map,
    addBreakpoint: Function,
    removeBreakpoint: Function,
    selectedSource: Object,
    selectedLocation: Object
  };

  constructor(props) {
    super(props);
    this.onKeyDown = this.onKeyDown.bind(this);
    this.onKeyUp = this.onKeyUp.bind(this);

    this.state = {
      showCallSites: false
    };
  }

  componentDidMount() {
    const { editor } = this.props;
    const codeMirrorWrapper = editor.codeMirror.getWrapperElement();

    if (features.columnBreakpoints) {
      codeMirrorWrapper.addEventListener("click", e => this.onTokenClick(e));
      document.body.addEventListener("keydown", this.onKeyDown);
      document.body.addEventListener("keyup", this.onKeyUp);
    }
  }

  componentDidUnMount() {
    const { editor } = this.props;
    const codeMirrorWrapper = editor.codeMirror.getWrapperElement();

    if (features.columnBreakpoints) {
      codeMirrorWrapper.addEventListener("click", e => this.onTokenClick(e));
      document.body.removeEventListener("keydown", e => this.onKeyDown);
      document.body.removeEventListener("keyup", this.onKeyUp);
    }
  }

  onKeyUp(e) {
    if (e.key === "Alt") {
      e.preventDefault();
      this.setState({ showCallSites: false });
    }
  }

  onKeyDown(e) {
    if (e.key === "Alt") {
      e.preventDefault();
      this.setState({ showCallSites: true });
    }
  }

  onTokenClick(e) {
    const { target } = e;
    const { editor, selectedLocation } = this.props;

    if (
      (!e.altKey && !target.classList.contains("call-site-bp")) ||
      (!target.classList.contains("call-site") &&
        !target.classList.contains("call-site-bp"))
    ) {
      return;
    }

    const { sourceId } = selectedLocation;
    const { line, column } = getTokenLocation(editor.codeMirror, target);

    this.toggleBreakpoint(line, isWasm(sourceId) ? undefined : column);
  }

  toggleBreakpoint(line, column = undefined) {
    const {
      selectedSource,
      selectedLocation,
      addBreakpoint,
      removeBreakpoint,
      callSites
    } = this.props;

    const callSite = getCallSiteAtLocation(callSites, { line, column });

    if (!callSite) {
      return;
    }

    const bp = callSite.breakpoint;

    if ((bp && bp.loading) || !selectedLocation || !selectedSource) {
      return;
    }

    const { sourceId } = selectedLocation;

    if (bp) {
      // NOTE: it's possible the breakpoint has slid to a column
      column = column || bp.location.column;
      removeBreakpoint({
        sourceId: sourceId,
        line: line,
        column
      });
    } else {
      addBreakpoint({
        sourceId: sourceId,
        sourceUrl: selectedSource.get("url"),
        line: line,
        column: column
      });
    }
  }

  render() {
    const { editor, callSites, selectedSource } = this.props;
    const { showCallSites } = this.state;
    let sites;
    if (!callSites) {
      return null;
    }

    editor.codeMirror.operation(() => {
      const childCallSites = callSites.map((callSite, index) => {
        const props = {
          key: index,
          callSite,
          editor,
          source: selectedSource,
          breakpoint: callSite.breakpoint,
          showCallSite: showCallSites
        };
        return <CallSite {...props} />;
      });
      sites = <div>{childCallSites}</div>;
    });
    return sites;
  }
}

function getCallSites(symbols, breakpoints) {
  if (!symbols || !symbols.callExpressions) {
    return;
  }

  const callSites = symbols.callExpressions;

  // NOTE: we create a breakpoint map keyed on location
  // to speed up the lookups. Hopefully we'll fix the
  // inconsistency with column offsets so that we can expect
  // a breakpoint to be added at the beginning of a call expression.
  const bpLocationMap = keyBy(breakpoints.valueSeq().toJS(), ({ location }) =>
    locationKey(location)
  );

  function locationKey({ line, column }) {
    return `${line}/${column}`;
  }

  function findBreakpoint(callSite) {
    const { location: { start, end } } = callSite;

    const breakpointId = range(start.column - 1, end.column)
      .map(column => locationKey({ line: start.line, column }))
      .find(key => bpLocationMap[key]);

    if (breakpointId) {
      return bpLocationMap[breakpointId];
    }
  }

  return callSites
    .filter(({ location }) => location.start.line === location.end.line)
    .map(callSite => ({ ...callSite, breakpoint: findBreakpoint(callSite) }));
}

const { addBreakpoint, removeBreakpoint } = actions;

export default connect(
  state => {
    const selectedLocation = getSelectedLocation(state);
    const selectedSource = getSelectedSource(state);
    const sourceId = selectedLocation && selectedLocation.sourceId;
    const source = selectedSource && selectedSource.toJS();

    const symbols = getSymbols(state, source);
    const breakpoints = getBreakpointsForSource(state, sourceId);

    return {
      selectedLocation,
      selectedSource,
      callSites: getCallSites(symbols, breakpoints),
      breakpoints: breakpoints
    };
  },
  {
    addBreakpoint,
    removeBreakpoint
  }
)(CallSites);
