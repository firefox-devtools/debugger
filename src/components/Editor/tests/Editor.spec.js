/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at <http://mozilla.org/MPL/2.0/>. */

import React from "react";
import { shallow } from "enzyme";
import Editor from "../index";

function generateDefaults(overrides) {
  return {
    toggleBreakpoint: jest.fn(),
    toggleDisabledBreakpoint: jest.fn(),
    addOrToggleDisabledBreakpoint: jest.fn(),
    ...overrides
  };
}

function createMockEditor() {
  return {
    codeMirror: {
      doc: {},
      setOption: jest.fn(),
      scrollTo: jest.fn(),
      charCoords: ({ line, ch }) => ({ top: line, left: ch }),
      getScrollerElement: () => ({ offsetWidth: 0, offsetHeight: 0 }),
      getScrollInfo: () => ({
        top: 0,
        left: 0,
        clientWidth: 0,
        clientHeight: 0
      }),
      defaultCharWidth: () => 0,
      defaultTextHeight: () => 0,
      display: { gutters: { querySelector: jest.fn() } }
    },
    setText: jest.fn(),
    on: jest.fn(),
    off: jest.fn(),
    createDocument: () => ({
      getLine: line => ""
    }),
    replaceDocument: jest.fn(),
    setMode: jest.fn()
  };
}

function createMockSource(overrides) {
  return {
    id: "foo",
    text: "the text",
    loadedState: "loaded",
    url: "foo",
    ...overrides
  };
}

function render(overrides = {}) {
  const props = generateDefaults(overrides);
  const mockEditor = createMockEditor();
  const component = shallow(<Editor.WrappedComponent {...props} />, {
    context: {
      shortcuts: { on: jest.fn() }
    },
    disableLifecycleMethods: true
  });

  return { component, props, mockEditor };
}

describe("Editor", () => {
  describe("When empty", () => {
    it("should render", async () => {
      const { component } = render();
      expect(component).toMatchSnapshot();
    });
  });

  describe("When loading initial source", () => {
    it("should show a loading message", async () => {
      const { component, mockEditor } = render();
      await component.setState({ editor: mockEditor });
      component.setProps({
        selectedSource: { loadedState: "loading" }
      });

      expect(mockEditor.setText.mock.calls).toEqual([["Loading…"]]);
      expect(mockEditor.codeMirror.scrollTo.mock.calls).toEqual([]);
    });
  });

  describe("When loaded", () => {
    it("should show text", async () => {
      const { component, mockEditor, props } = render({});

      await component.setState({ editor: mockEditor });
      await component.setProps({
        ...props,
        selectedSource: createMockSource({ loadedState: "loaded" }),
        selectedLocation: { sourceId: "foo", line: 3, column: 1 }
      });

      expect(mockEditor.setText.mock.calls).toEqual([["the text"]]);
      expect(mockEditor.codeMirror.scrollTo.mock.calls).toEqual([[1, 2]]);
    });
  });

  describe("When error", () => {
    it("should show error text", async () => {
      const { component, mockEditor, props } = render({});

      await component.setState({ editor: mockEditor });
      await component.setProps({
        ...props,
        selectedSource: createMockSource({
          loadedState: "loaded",
          text: undefined,
          error: "error text"
        }),
        selectedLocation: { sourceId: "bad-foo", line: 3, column: 1 }
      });

      expect(mockEditor.setText.mock.calls).toEqual([
        ["Error loading this URI: error text"]
      ]);
    });

    it("should show wasm error", async () => {
      const { component, mockEditor, props } = render({});

      await component.setState({ editor: mockEditor });
      await component.setProps({
        ...props,
        selectedSource: createMockSource({
          loadedState: "loaded",
          isWasm: true,
          text: undefined,
          error: "blah WebAssembly binary source is not available blah"
        }),
        selectedLocation: { sourceId: "bad-foo", line: 3, column: 1 }
      });

      expect(mockEditor.setText.mock.calls).toEqual([
        ["Please refresh to debug this module"]
      ]);
    });
  });

  describe("When navigating to a loading source", () => {
    it("should show loading message and not scroll", async () => {
      const { component, mockEditor, props } = render({});

      await component.setState({ editor: mockEditor });
      await component.setProps({
        ...props,
        selectedSource: createMockSource({ loadedState: "loaded" }),
        selectedLocation: { sourceId: "foo", line: 3, column: 1 }
      });

      // navigate to a new source that is still loading
      await component.setProps({
        ...props,
        selectedSource: createMockSource({
          id: "bar",
          loadedState: "loading"
        }),
        selectedLocation: { sourceId: "bar", line: 1, column: 1 }
      });

      expect(mockEditor.setText.mock.calls).toEqual([
        ["the text"],
        ["Loading…"]
      ]);

      expect(mockEditor.codeMirror.scrollTo.mock.calls).toEqual([[1, 2]]);
    });
  });

  describe("When navigating to a loaded source", () => {
    it("should show text and then scroll", async () => {
      const { component, mockEditor, props } = render({});

      await component.setState({ editor: mockEditor });
      await component.setProps({
        ...props,
        selectedSource: createMockSource({ loadedState: "loading" }),
        selectedLocation: { sourceId: "foo", line: 1, column: 1 }
      });

      // navigate to a new source that is still loading
      await component.setProps({
        ...props,
        selectedSource: createMockSource({
          loadedState: "loaded"
        }),
        selectedLocation: { sourceId: "foo", line: 1, column: 1 }
      });

      expect(mockEditor.setText.mock.calls).toEqual([
        ["Loading…"],
        ["the text"]
      ]);

      expect(mockEditor.codeMirror.scrollTo.mock.calls).toEqual([[1, 0]]);
    });
  });
});
