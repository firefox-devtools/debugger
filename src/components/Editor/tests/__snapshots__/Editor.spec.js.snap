// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Editor should render 1`] = `
ShallowWrapper {
  "complexSelector": ComplexSelector {
    "buildPredicate": [Function],
    "childrenOfNode": [Function],
    "findWhereUnwrapped": [Function],
  },
  "length": 1,
  "node": <div
    className="editor-wrapper"
>
    <div
        className="editor-mount devtools-monospace"
        style={
            Object {
                "height": "calc(100% - var(--editor-footer-height))",
              }
        }
    />
</div>,
  "nodes": Array [
    <div
      className="editor-wrapper"
>
      <div
            className="editor-mount devtools-monospace"
            style={
                  Object {
                        "height": "calc(100% - var(--editor-footer-height))",
                      }
            }
      />
</div>,
  ],
  "options": Object {},
  "renderer": ReactShallowRenderer {
    "_instance": ShallowComponentWrapper {
      "_calledComponentWillUnmount": false,
      "_compositeType": 1,
      "_context": Object {},
      "_currentElement": <Editor
        addBreakpoint={[Function]}
        addExpression={[Function]}
        addOrToggleDisabledBreakpoint={[Function]}
        breakpoints={
                Immutable.Map {
                }
        }
        clearPreview={[Function]}
        disableBreakpoint={[Function]}
        enableBreakpoint={[Function]}
        getExpression={[Function]}
        query=""
        removeBreakpoint={[Function]}
        searchModifiers={
                Object {
                        "caseSensitive": false,
                        "regexMatch": false,
                        "wholeWord": false,
                      }
        }
        setBreakpointCondition={[Function]}
        toggleBreakpoint={[Function]}
        toggleConditionalBreakpointPanel={[Function]}
        toggleDisabledBreakpoint={[Function]}
/>,
      "_debugID": 1,
      "_hostContainerInfo": null,
      "_hostParent": null,
      "_instance": Editor {
        "_reactInternalInstance": [Circular],
        "cbPanel": null,
        "closeConditionalPanel": [Function],
        "context": Object {
          "shortcuts": undefined,
        },
        "lastJumpLine": null,
        "onEscape": [Function],
        "onGutterClick": [Function],
        "onGutterContextMenu": [Function],
        "onSearchAgain": [Function],
        "onToggleBreakpoint": [Function],
        "pendingJumpLocation": null,
        "props": Object {
          "addBreakpoint": [Function],
          "addExpression": [Function],
          "addOrToggleDisabledBreakpoint": [Function],
          "breakpoints": Immutable.Map {
},
          "clearPreview": [Function],
          "disableBreakpoint": [Function],
          "enableBreakpoint": [Function],
          "getExpression": [Function],
          "query": "",
          "removeBreakpoint": [Function],
          "searchModifiers": Object {
            "caseSensitive": false,
            "regexMatch": false,
            "wholeWord": false,
          },
          "setBreakpointCondition": [Function],
          "toggleBreakpoint": [Function],
          "toggleConditionalBreakpointPanel": [Function],
          "toggleDisabledBreakpoint": [Function],
        },
        "refs": Object {},
        "state": Object {
          "editor": null,
          "highlightedLineRange": null,
        },
        "toggleConditionalPanel": [Function],
        "updater": Object {
          "enqueueCallback": [Function],
          "enqueueCallbackInternal": [Function],
          "enqueueElementInternal": [Function],
          "enqueueForceUpdate": [Function],
          "enqueueReplaceState": [Function],
          "enqueueSetState": [Function],
          "isMounted": [Function],
          "validateCallback": [Function],
        },
      },
      "_mountOrder": 4,
      "_pendingCallbacks": null,
      "_pendingElement": null,
      "_pendingForceUpdate": false,
      "_pendingReplaceState": false,
      "_pendingStateQueue": null,
      "_renderedComponent": Object {
        "_currentElement": <div
          className="editor-wrapper"
>
          <div
                    className="editor-mount devtools-monospace"
                    style={
                              Object {
                                        "height": "calc(100% - var(--editor-footer-height))",
                                      }
                    }
          />
</div>,
        "_debugID": 2,
        "_renderedOutput": <div
          className="editor-wrapper"
>
          <div
                    className="editor-mount devtools-monospace"
                    style={
                              Object {
                                        "height": "calc(100% - var(--editor-footer-height))",
                                      }
                    }
          />
</div>,
      },
      "_renderedNodeType": 0,
      "_rootNodeID": 0,
      "_topLevelWrapper": null,
      "_updateBatchNumber": null,
      "_warnedAboutRefsInRender": false,
    },
    "getRenderOutput": [Function],
    "render": [Function],
  },
  "root": [Circular],
  "unrendered": <Editor
    addBreakpoint={[Function]}
    addExpression={[Function]}
    addOrToggleDisabledBreakpoint={[Function]}
    breakpoints={
        Immutable.Map {
        }
    }
    clearPreview={[Function]}
    disableBreakpoint={[Function]}
    enableBreakpoint={[Function]}
    getExpression={[Function]}
    query=""
    removeBreakpoint={[Function]}
    searchModifiers={
        Object {
            "caseSensitive": false,
            "regexMatch": false,
            "wholeWord": false,
          }
    }
    setBreakpointCondition={[Function]}
    toggleBreakpoint={[Function]}
    toggleConditionalBreakpointPanel={[Function]}
    toggleDisabledBreakpoint={[Function]}
/>,
}
`;
